### PR TITLE
DAOS-3620 control: combine nvme structs at source

### PR DESCRIPTION
**Pull request by tanabarr**
Aggregate health statistics and namespace structures with controller at
bindings rather than later in the discovery flow in order to simplify 
translation. Add cmocka tests to provide some coverage of spdk bindings 
and to verify this aggregation.

Signed-off-by: Amanda Justiniano amanda.justiniano-pagn@intel.com
Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/lib/spdk/nvme.go [hunk 1]
```diff
@@ -215,66 +208,35 @@ func processReturn(retPtr *C.struct_ret_t, failLocation string) (
...
-				continue
+			if healthPtr != nil {
+				defer C.free(unsafe.Pointer(healthPtr))
+				ctrlrRef.Health = c2GoDeviceHealth(healthPtr)
 			}
 
-			defer C.free(unsafe.Pointer(healthPtr))
-			devs = append(devs, c2GoDeviceHealth(ctrlr.PCIAddr, healthPtr))
+			// store reference in map so we can add amespaces later
+			pciMap[ctrlrRef.PCIAddr] = ctrlrRef
```
Comment (mjmac) [inline src/control/lib/spdk/nvme.go]: Should there be a check first to see if there's already a reference in the map?

File: src/control/lib/spdk/nvme_test.go [hunk 2]
```diff
@@ -24,9 +24,15 @@
 package spdk
 
 import (
+	"C"
```
Comment (mjmac) [inline src/control/lib/spdk/nvme_test.go]: Pretty sure you can't import C in _test.go files.

File: src/control/lib/spdk/nvme_test.go [hunk 3]
```diff
@@ -24,9 +24,15 @@
...
 	"fmt"
 	"testing"
 )
+import (
+	"github.com/daos-stack/daos/src/control/common"
+	"github.com/daos-stack/daos/src/control/logging"
+	"github.com/google/go-cmp/cmp"
```
Comment (mjmac) [inline src/control/lib/spdk/nvme_test.go]: Misplaced import.

**Comment by tanabarr** [comment]
building and running cmocka tests (which needs to be integrated intoI CI)
```
tanabarr@boro-45:~/projects/daos_m/src/control/lib/spdk/ctests$  gcc ${TESTFLAGS} ${DISCOVER_WRAPS} -I../include -L../../../../../install/lib nvme_discover_test.c ../src/*.c -o discover_ctest
tanabarr@boro-45:~/projects/daos_m/src/control/lib/spdk/ctests$ ./discover_ctest                                                                                [==========] Running 3 test(s).
[ RUN      ] test_nvme_discover_null_g_controllers
no NVMe controllers found
Mock cleanup()...
nvme_discover() returned: 15003776
[       OK ] test_nvme_discover_null_g_controllers
[ RUN      ] test_nvme_discover_set_g_controllers
no NVMe controllers found
Mock cleanup()...
nvme_discover() returned: 15003776
[       OK ] test_nvme_discover_set_g_controllers
[ RUN      ] test_nvme_discover_nvme_probe_fail
spdk_nvme_probe() failed
Mock cleanup()...
nvme_discover() returned: 15003776
[       OK ] test_nvme_discover_nvme_probe_fail
[==========] 3 test(s) run.
[  PASSED  ] 3 test(s).
```

File: src/control/lib/spdk/ctests/nvme_discover_test.c [hunk 4]
```diff
@@ -0,0 +1,175 @@
+/**
```
Comment (kjacque) [inline src/control/lib/spdk/ctests/nvme_discover_test.c]: Any reason not to build/run these tests for CI?

Comment (tanabarr) [reply src/control/lib/spdk/ctests/nvme_discover_test.c]: just haven't finished the PR yet

File: src/control/lib/spdk/ctests/nvme_discover_test.c [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+		g_controllers->ctrlr = NULL;
+		g_controllers->dev_health = NULL;
+		g_controllers->next = NULL;
+	}
+
+	/* Define param checks and return values */
+	will_return(__wrap_init_ret, &ret);
+	will_return(__wrap_spdk_nvme_probe, 0);
+	expect_any(__wrap_spdk_nvme_probe, probe_cb);
+	expect_any(__wrap_spdk_nvme_probe, attach_cb);
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/lib/spdk/ctests/nvme_discover_test.c]: not sure

**Comment by tanabarr** [comment]
replaced by https://github.com/daos-stack/daos/pull/1479

**Pull request closed without merging**